internal/cloud/azure/azure.go

```diff
@@ -388,6 +446,89 @@ func (c *Cloud) getLoadBalancerDNSName(ctx context.Context) (string, error) {
 }
 */
 
+// PrepareAzureControlPlaneNode sets up iptables for the control plane node only
+// if an internal load balancer is used.
+//
+// This is needed since during `kubeadm init` the API server must talk to the
+// kubeAPIEndpoint, which is the load balancer IP address. During that time, the
+// only healthy VM is the VM itself. Therefore, traffic is sent to the load balancer
+// and the 5-tuple is (VM IP, <some port>, LB IP, 6443, TCP).
+// Now the load balancer does not re-write the source IP address only the destination (DNAT).
+// Therefore the 5-tuple is (VM IP, <some port>, VM IP, 6443, TCP).
+// Now the VM responds to the SYN packet with a SYN-ACK packet, but the outgoing
+// connection waits on a response from the load balancer and not the VM therefore
+// dropping the packet.
+//
+// OpenShift also uses the same mechanism to redirect traffic to the API server:
+// https://github.com/openshift/machine-config-operator/blob/e453bd20bac0e48afa74e9a27665abaf454d93cd/templates/master/00-master/azure/files/opt-libexec-openshift-azure-routes-sh.yaml
+func (c *Cloud) PrepareAzureControlPlaneNode(ctx context.Context, log *logger.Logger) {
+	chainName := "azure-lb-nat"
+
+	selfMetadata, err := c.Self(ctx)
+	if err != nil {
+		log.Fatalf("failed to get self metadata: %v", err)
+	}
+
+	if selfMetadata.Role != role.ControlPlane {
+		log.Infof("not a control plane node, skipping iptables setup")
+		return
+	}
+
+	loadbalancerIP, err := c.getLoadBalancerPrivateIP(ctx)
+	if err != nil {
+		log.Infof("failed to get load balancer private IP: %v", err)
+		return
+	}
+
+	log.Infof("Setting up iptables for control plane node with load balancer IP %s", loadbalancerIP)
+
+	// Wait for the API server to be ready
+	httpClient := http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		},
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", "https://localhost:6443/readyz", nil)
+	if err != nil {
+		log.Fatalf("failed to create request: %v", err)
+	}
+
+	for {
+		time.Sleep(5 * time.Second)
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			fmt.Printf("failed to request endpoint: %v\n", err)
+			continue
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode == http.StatusOK {
+			break
+		}
+	}
+
+	// Create the iptables chain
+	out, err := exec.CommandContext(ctx, "iptables", "-w", "-t", "nat", "-N", chainName).CombinedOutput()
```

Comment:
Have you considered using [`iptables` from CoreOS](https://pkg.go.dev/github.com/coreos/go-iptables/iptables)?
In the end this is still just a wrapper for the iptables binary, but gives a nicer interface.

Reply:
Yes, I had a look at wrapping the commands but decided against it since it seemed like that we'd only need a very small part of the package. But I have no strong opinions here. I'll implement implement it with the package and see how it looks.

Reply:
Looks great, using it.